Update security model of Airflow (#32098)
mistercrunch/Airflow: docs/apache-airflow/security/index.rst

 .. Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

 ..   http://www.apache.org/licenses/LICENSE-2.0

 .. Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

Security
========

.. toctree::
    :maxdepth: 1
    :glob:

    *
    secrets/index

This document describes Airflow's security model from the perspective of
the Airflow user. It is intended to help users understand the security
model and make informed decisions about how to deploy and manage Airflow.

> If you would like to know, how to report security vulnerabilities and how

Reviewer: ```suggestion
If you would like to know how to report security vulnerabilities and how
```

Author: yep

security reports are handled by the security team of Airflow, head to
`Airflow's Security Policy <https://github.com/apache/airflow/security/policy>`_.

Airflow Security Model - user types
-----------------------------------

The Airflow security model involves different types of users with
varying access and capabilities:

1. **Deployment Managers**: They have the highest level of access and
   control. They install and configure Airflow, make decisions about
   technologies and permissions. They can potentially delete the entire
   installation and have access to all credentials. Deployment Managers
   can also decide to keep audits, backups and copies of information
   outside of Airflow, which are not covered by Airflow's security
   model.

2. **DAG Authors**: They can upload, modify, and delete DAG files,

>    running code and accessing credentials. They have full access to the

Reviewer: Could we clarify this a bit more, please? Do we mean "They can upload, modify, and delete DAG files, execute code and access credentials"? or "They can upload, modify and delete DAG files that execute code and access credentials"?

Author: Yeah. I added a bit longer and more precise description.

   metadata database and internal audit logs.

> 3. **Authenticated UI users**: They have access to the UI and API. Admin

Reviewer: Should we add an initial statement here something like "Broadly, there are three types of authenticated UI users, namely Admin users, Connection configuration users, and Operations users." and then go about explaining what they mean?

Author: I had it like that initially but it feels quite repetitive. I limited myself (original description was quite a bit longer and while my natural way would be similar, I think we do not loose much without this "preamble".

    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

 ..   http://www.apache.org/licenses/LICENSE-2.0

 .. Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

Security
========

.. toctree::
    :maxdepth: 1
    :glob:

    *
    secrets/index

This document describes Airflow's security model from the perspective of
the Airflow user. It is intended to help users understand the security
model and make informed decisions about how to deploy and manage Airflow.
If you would like to know how to report security vulnerabilities and how
security reports are handled by the security team of Airflow, head to
`Airflow's Security Policy <https://github.com/apache/airflow/security/policy>`_.

> Airflow Security Model - user types

Reviewer 2: in the .github/security.rst titles only had the first word capitalized (other than proper nouns of course) so we should probably follow the same in this one?:

```suggestion
Airflow security model - user types
```

-----------------------------------

The Airflow security model involves different types of users with
varying access and capabilities:

1. **Deployment Managers**: They have the highest level of access and
   control. They install and configure Airflow, make decisions about
   technologies and permissions. They can potentially delete the entire
   installation and have access to all credentials. Deployment Managers
   can also decide to keep audits, backups and copies of information
   outside of Airflow, which are not covered by Airflow's security
   model.

2. **DAG Authors**: They can upload, modify, and delete DAG files. The
   code in DAG files is executed on workers. Therefore, DAG authors can create
   and change code executed on workers and potentially access the credentials
   that DAG code uses to access external systems. DAG Authors have full access

>    to the metadata database and internal audit logs.

Reviewer 2: Should we note that DB access is planned to change soon with the multi-tenancy work underway?

Author: I am hesitating a bit to add "future" plans. I had it initially and AIPs mentioned, but I thought, well what's not done, is not done (yet) and we should not really mention it untitl it is ready for people to use (even in experimental mode).

Reviewer 2: Fair enough, I see that argument. As long as we remember to circle back and update this doc :+1:

Author: I will :reminder_ribbon:

3. **Authenticated UI users**: They have access to the UI and API. Admin
   users can manage permissions and execute code on workers. Connection
   configuration users can configure connections and execute code on
   workers. Operations users have access to DAG execution status. Trust
   is crucial to prevent abuse and Denial of Service attacks.

4. **Non-authenticated UI users**: Airflow doesn't support
   unauthenticated users by default. If allowed, vulnerabilities must be
   addressed by the Deployment Manager.

> Capabilities of Authenticated UI users

Reviewer 2: ```suggestion
Capabilities of authenticated UI users
```

--------------------------------------

The capabilities of **Authenticated UI users** can vary depending on

> whether the users belong to one of the general categories:

Reviewer 2: I'd like to call out that Roles and Permissions in Airflow are fluid and configurable, so the below three roles are not exhaustive, but are still helpful. Here's a proposed paragraph to lead into that list: 
```suggestion
The capabilities of **Authenticated UI users** can vary depending on
what roles have been configured by the Deployment Manager or Admin users as well as what permissions those roles have. Permissions on roles can be scoped as tightly as a single DAG, for example, or as broad as Admin. Below are three general categories to help conceptualize some of the capabilities authenticated users may have:
```

Author: I like this one :)

   server and cause Denial of Service situations.

Responsibilities of Deployment Managers
---------------------------------------

Deployment Managers determine access levels and understand the potential
damage users can cause. Some Deployment Managers may further limit
access through fine-grained privileges for the **Authenticated UI
users**. However, these limitations are outside the basic Airflow's
security model and are at the discretion of Deployment Managers.
Examples of fine-grained access control include (but are not limited
to):

-  Limiting login permissions: Restricting the accounts that users can
   log in with, allowing only specific accounts or roles belonging to
   access the Airflow system.

-  Access restrictions to views or DAGs: Controlling user access to
   certain views or specific DAGs, ensuring that users can only view or
   interact with authorized components.

-  Implementing static code analysis and code review: Introducing
   processes such as static code analysis and code review as part of the
   DAG submission pipeline. This helps enforce code quality standards,
   security checks, and adherence to best practices before DAGs are

>    deployed.

Reviewer 2: Perhaps out of scope for a document focused on Airflow itself? I don't feel strongly though, happy to keep it.

Author: I think it adds value. I think the main point I have here is that Airlfow is "inherently insecure" because DAG authors can do everything. And this is both yes and no. It's kind of symilar to code submission to Software Repository - anyone can submit a code, but we have a gate in form of review of the code before it is submitted and it's better than anything you could do automatically I think (if you treat it seriously).

It is part of the general lack of understanding that you can apply standard engineering practices to your DAG pipeline and that can bring one level of security that might be "good enough" for many of our users.  And it also provides extensibility of the security model - as for example one could have already a very simple multi-tenanti implementation of Airlfow if they add code review and automated CI checks to the mix of things. 

I had this many times when I talked to the users - and when they already started to doubt into their "split teams"  idea , I thrown that one in the mix and they were " ah yaeah, indeed that might work".

These examples showcase ways in which administrators can refine and
limit user privileges within Airflow, providing tighter control and
ensuring that users have access only to the necessary components and
functionalities based on their roles and responsibilities, however the
fine-grained access control does not provide full isolation and
separation of access allowing isolation of different user groups in a
multi-tenant fashion yet. In the future versions of Airflow some of the
fine-grained access features might become part of the Airflow security

> model. Airflow community works on a multi-tenant model that might
> address some of the fine-grained access components.

Reviewer 2: ```suggestion
model. The Airflow community is working on a multi-tenant model that might
address some of the fine-grained access components.
```

Releasing Airflow with security patches
---------------------------------------

> Apache Airflow uses strict [SemVer](https://semver.org) versioning policy, which means that we strive for

Reviewer 2: ```suggestion
Apache Airflow uses a strict [SemVer](https://semver.org) versioning policy, which means that we strive for
```

any release of a given ``MAJOR`` Version (version "2" currently) to be backwards compatible. When we

> release ``MINOR`` version, the development continues in the ``main`` branch where we prepare the next

Reviewer 2: ```suggestion
release a ``MINOR`` version, the development continues in the ``main`` branch where we prepare the next
```

``MINOR`` version, but we release ``PATCHLEVEL`` releases with selected bugfixes (including security
bugfixes) cherry-picked to the latest released ``MINOR`` line of Apache Airflow. At the moment, when we
release a new ``MINOR`` version, we stop releasing ``PATCHLEVEL`` releases for the previous ``MINOR`` version.

> For example, when we released  ``2.6.0`` version on April 30, 2023, until we release ``2.7.0`` version,
> all the security patches will be cherry-picked and released in ``2.6.*`` versions only. There will be no
> ``2.5.*`` versions  released after ``2.6.0`` has been released.

Reviewer 2: ```suggestion
For example, once we released ``2.6.0`` version on April 30, 2023 all the security patches will be cherry-picked and released in ``2.6.*`` versions until we release ``2.7.0`` version. There will be no
``2.5.*`` versions  released after ``2.6.0`` has been released.
```

This means that in order to apply security fixes with Apache Airflow software released by us, you
MUST upgrade to the latest ``MINOR`` version of Airflow.

> Releasing Airflow Providers with security patches

Reviewer 2: ```suggestion
Releasing Airflow providers with security patches
```
